### PR TITLE
Implement neg for all types

### LINKED ISSUE #4066: Negative mask unsupport (ByteTensor) in Pytorch 0.3
When I use the negative mask like : 

`mask = torch.autograd.Variable(torch.zeros((3,4 ))).byte()`
`mask =  1- mask `
Pytorch 0.3 shows "RuntimeError: neg is not implemented for type torch.ByteTensor"

It is also observed in Pytorch 0.2, but can be solved in Pytorch 0.2 using:
`mask =  1 + (-1) *mask`

While in Pytorch 0.3, the previous solution gives the "RuntimeError: value cannot be converted to type uint8_t without overflow: -1".

It is caused by the ByteTensor does not support negative operation. So I use  `mask = (1-mask.long()).byte()` to solve the problem, but it is not elegant.

### PR DESCRIPTION
**Pull request by colesbury**
The C/C++ unary negation operator is well defined for unsigned types. We
should use that behavior. This also implements neg for CharTensor. That
behavior currently depends on whether char is signed or unsigned.

Fixes #4066, #3225

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -525,7 +525,7 @@ def test_csub(self):
     @staticmethod
     def _test_neg(self, cast):
         float_types = ['torch.DoubleTensor', 'torch.FloatTensor', 'torch.LongTensor']
-        int_types = ['torch.IntTensor', 'torch.ShortTensor']
```
Comment (gchanan) [inline test/test_torch.py]: test for CharTensor?  I don't know if there's an easy way from python to figure out the what the behavior should be.

Comment (colesbury) [reply test/test_torch.py]: Done

**Review by gchanan — APPROVED** [review]
lgtm.

**Comment by pietern** [comment]
There was some CI maintenance happening this morning -- retriggering build.

@pytorchbot retest this please

**Pull request merged**